src/server/qtquick/wsurfaceitem.cpp

```diff
@@ -256,12 +257,13 @@ class WSurfaceItemContentPrivate : public WObjectPrivate
         if (frameDoneConnection)
             QObject::disconnect(frameDoneConnection);
 
-        if (!q_func()->isVisible())
-            return;
-
         if (auto output = surface->primaryOutput()) {
-            frameDoneConnection = QObject::connect(output, &WOutput::bufferCommitted,
-                                                   surface, &WSurface::notifyFrameDone);
+            frameDoneConnection = QObject::connect(output, &WOutput::bufferCommitted,[&](){
```

Comment:
Don't connect to `WOutput::bufferCommitted`. Should notify frame done after render, see QQuickWindow::afterRendering/QQuickWindow::AfterRenderingStage.

We can call `notifyFrameDone` on WSGRenderFootprintNode::render, for an example:

````
void render(const RenderState*) override
{
        owner->window()->scheduleRenderJob(..., QQuickWindow::AfterRenderingStage);
}
````

Reply:
agree but don't know would texture be changed before qsg's commands are done on hardware

Reply:
> Don't connect to `WOutput::bufferCommitted`. Should notify frame done after render, see QQuickWindow::afterRendering/QQuickWindow::AfterRenderingStage.
> 
> We can call `notifyFrameDone` on WSGRenderFootprintNode::render, for an example:
> 
> ```
> void render(const RenderState*) override
> {
>         owner->window()->scheduleRenderJob(..., QQuickWindow::AfterRenderingStage);
> }
> ```

The afterRendering signal is used, imo notifyFrameDone should not be called from rendering thrd (currently seems rendering thrd is not seperated)